crypto/ec/ec_local.h

```diff
@@ -213,6 +213,8 @@ struct ec_group_st {
     BIGNUM *order, *cofactor;
     int curve_name;             /* optional NID for named curve */
     int asn1_flag;              /* flag to control the asn1 encoding */
+    int decoded_from_explicit_params; /* set if decoded from explicit
+                                       * curve parameters encoding */
```

Comment:
Actually this and the new public function seems not needed if you generalize the check because RFC 5480 forbids also implicitCurve:
`check_curve()` could simply check if  `curve_name` contains a valid NID, which would imply not implicit and not explicit parameters, right?

Reply:
Nope, that would not work as the curve name is set also for keys that have explicit params that matches one of the named curves. And as the implicit parameters are not supported at all, they do not need to be flagged this way. I'd like to keep this as is.

Reply:
I see

Reply:
> that would not work as the curve name is set also for keys that have explicit params that matches one of the named curves. And as the implicit parameters are not supported at all, they do not need to be flagged this way. 

Hmm, since this PR is still under discussion I'd like to revisit this point.
If a named curve happens to match a predefined curve I'd say this does not need to be flagged as wrong because the essence of the RFC requirement is that 'custom` curves must not be used.
And I tentatively did the following simple change:
```
diff --git a/crypto/x509/x509_vfy.c b/crypto/x509/x509_vfy.c
index b7afdbbccb..046518d8ae 100644
--- a/crypto/x509/x509_vfy.c
+++ b/crypto/x509/x509_vfy.c
@@ -3460,7 +3460,7 @@ static int check_curve(X509 *cert)
         return 1;
 
     if (EVP_PKEY_id(pkey) == EVP_PKEY_EC
-        && EC_KEY_decoded_from_explicit_params(EVP_PKEY_get0_EC_KEY(pkey)))
+        && EC_GROUP_get_curve_name(EC_KEY_get0_group(EVP_PKEY_get0_EC_KEY(pkey))) != NID_undef)
         return 0;
 #endif
 
```
and voila, the tests still pass.
This way the function `EC_KEY_decoded_from_explicit_params()` is not really needed, as I suspected.

Reply:
RFC5480 says: `implicitCurve and specifiedCurve MUST NOT be used in PKIX.` So the only option is that namedCurve is used.

Reply:
So the question is on how strict we want to enforce that. The CABForum has in it's BRs: "The parameters MUST use the namedCurve encoding.", and then exact bytes for each of the allowed curves.

I have yet to find the first real certificate that doesn't use the namedCurve. The only ones I've seen have been converted from namedCurve and then fail the signature check, or test certificates to show what should get rejected.

Reply:
> RFC5480 says: implicitCurve and specifiedCurve MUST NOT be used in PKIX. So the only option is that namedCurve is used.

Yes, that's why I think also implicit curves should be rejected here.
And I'd subsume any cert with explicit params that coincide with a named curve as having that named curve, since this is essentially equivalent (at least from the security standpoint).

Reply:
> I have yet to find the first real certificate that doesn't use the namedCurve. The only ones I've seen have been converted from namedCurve and then fail the signature check, or test certificates to show what should get rejected.

Is there any EC cert in our test data that has an implicit or explicit curve?
I could not find any such.

Reply:
See #12139

Reply:
So @t8m you could use the cert `outputCerts/ECROOTv3_CAv3CustomCurve_LEAF_ECv3__ca_certificate1.pem` given in the zip file there.

Reply:
> Yes, that's why I think also implicit curves should be rejected here.

The fact is that certificates with implicit curves cannot really be represented by the X509 structure as the embedded EVP_PKEY cannot be really created. So complicating the API contract of  the EC_KEY_decoded_from_explicit_params() is not IMO worth it because the implicit case really cannot happen. So no, I am not going to change it.

Reply:
I see the point that with implicit curve it would not be possible to create an EC key as long as it is unclear where else the concrete curve information should come from. Yet it must be possible to create a cert bearing an EC key where the AlgorithmIdentifier has OID id-ecPublicKey and parameters NULL (= implicitCurve). Otherwise the restriction in RFC 5480 that is forbidden for conforming certs would make no sense. @kroecks, can you provide such a cert for adding a rexpective test?

Reply:
Yes, RFC allows creating such cert, but does OpenSSL allow loading it? I.e. will d2i_X509() on such cert return valid X509?

Reply:
I'm not sure if OpenSSL would be able to load a cert with implictCurve, which is another reason I hope that @kroeckx has a test cert such that we can give it a try.

Reply:
I propose handling implicitCurve certs in another PR.

Reply:
So I have a test where I get nid == NID_X9_62_id_ecPublicKey && parameter != NULL && parameter->type != V_ASN1_OBJECT, which triggers on 7 certificates in the database, I think that's not what you're talking about. I also have a check for parameter == NULL, but I've just marked that as invalid. I have 55 certificates marked as invalid, and should probably look at why they are marked invalid and split it out in different errors. Most of the invalid cases in x509lint are where OpenSSL returns an error, and that's an exception, I should fix that.

Reply:
All the certs that were marked invalid are cases where it failed to parse either the EKU or the SKID, issued by 2 CAs.